cmd/ttn-lw-cli/commands/clients.go

```diff
@@ -30,6 +30,8 @@ import (
 var (
 	selectClientFlags = util.FieldMaskFlags(&ttnpb.Client{})
 	setClientFlags    = util.FieldFlags(&ttnpb.Client{})
+
+	forbiddenSelectClientFlags = []string{"secret"}
```

Comment:
It's not forbidden; 1) it's just paths that aren't fetched by default when `-all` is passed and 2) forbidden or not would be a server decision.

Reply:
It may indeed be good to filter the paths using `ttnpb.AllowedFieldMaskPathsForRPC["/ttn.lorawan.v3.ClientRegistry/Get"]`. We could add such functionality to `util.FieldMaskFlags`.

Reply:
Maybe this could fit better inside a GRPC unary interceptor? Get the field mask paths of the request, and remove any that do not exist in `ttnpb.AllowedFieldMaskPathsForRPC[method_name]`

Reply:
In any case, the problem we are trying to solve with this is the following:

```
$ ttn-lw-cli clients list --all
error:pkg/rpcmiddleware/validator:field_mask_paths (forbidden path(s) in field mask)
    forbidden_paths=[secret]
    correlation_id=85b624ee54fe4b2da6647f65fd4e16ba
exit status 255
```

The `all` flag selects some of the forbidden field mask paths by default. The current solution is to hard code the paths that need to be removed from the field mask paths, and it is similar to what we do for the `AllowedFieldMaskPathsForRPC` as well:

https://github.com/TheThingsNetwork/lorawan-stack/blob/6e25d130324f2486842019a6aaafc8c3e825e626/pkg/ttnpb/field_mask_validation.go#L94

Reply:
Another alternative could be to allow an `all` or `*` field mask on the server side, and pass that from the CLI (that way anyone that uses the API can benefit from this)

Reply:
> The current solution is to hard code the paths that need to be removed from the field mask paths, and it is similar to what we do for the `AllowedFieldMaskPathsForRPC` as well

OK I see, but can't you use this function then? Just before you make the call, you strip the paths. A unary client interceptor is not a bad idea, but it also feels like a work-around. I'll leave this up to @htdvisser 

> Another alternative could be to allow an `all` or `*` field mask on the server side, and pass that from the CLI (that way anyone that uses the API can benefit from this)

Nah, this is really client-side. The question is where to do the filtering; in the "application" layer (i.e. commands), or in the service layer (i.e. API client).

Reply:
I don't think this should be done in the "Application" layer, mostly because this would require patching every single API call across multiple files.

Using `AllowedFieldMaskPathsForRPC` is not really an option either, because we don't know the method name at that point (and hard-coding it is similarly a bad idea)

Reply:
So, if we are to use a GRPC interceptor, we would need to something like the following:

```

// filterFieldMaskPaths is an interceptor that filters field mask paths that are not allowed
func filterFieldMaskPaths(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
	switch req.(type) {
	case *ttnpb.GetGatewayRequest:
		req.(*ttnpb.GetGatewayRequest).FieldMask.Paths = ttnpb.AllowedFieldMaskPathsForRPC[method]
	case *ttnpb.ListGatewaysRequest:
		req.(*ttnpb.ListGatewaysRequest).FieldMask.Paths = ttnpb.AllowedFieldMaskPathsForRPC[method]
        // more here...
	default:
	}

	return invoker(ctx, method, req, reply, cc, opts...)
}
```

Reply:
I don't think a client-side gRPC interceptor is appropriate here. We're already constructing the field masks in each command, so it's more appropriate to also do the filtering in the commands. The `applications list` command knows that it's calling the `ApplicationRegistry.List` RPC, so it can easily filter the fields using the allowed fields for that RPC. This is very similar to what we already do in `end_devices_split.go`.

---

If we do choose to go with the client interceptor, we should make it generic, instead of type checking the request for known messages.

```go
func Intercept(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
	if v, ok := req.(interface {
		GetFieldMask() pbtypes.FieldMask
	}); ok {
		fm := v.GetFieldMask()
		if allowedPaths, ok := ttnpb.AllowedFieldMaskPathsForRPC[method]; ok {
			fm.Paths = ttnpb.AllowedFields(fm.Paths, allowedPaths)
		}
	}
	return invoker(ctx, method, req, reply, cc)
}
```

Reply:
For the record, this is what I initially thought to do as well. But it does not work, because `fm` is a copy of the request FieldMask. I didn't find any way to mutate the field mask of a request without explicitly using its type (hence the example above).

Reply:
For the commands that are affected (`clients` and `users`), the flags are now filtered using `ttnpb.AllowedFieldMaskPathsForRPC`